### PR TITLE
Collect statistics for AddExchanges

### PR DESCRIPTION
**Pull request by kokosing**
Collect statistics for AddExchanges

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/StatsRecorder.java [hunk 1]
```diff
@@ -1,79 +0,0 @@
...
-import java.util.HashMap;
-import java.util.Map;
-
-import static com.google.common.base.Preconditions.checkArgument;
-import static com.google.common.base.Verify.verify;
-import static java.lang.String.format;
-
-public class StatsRecorder
-{
-    private final Map<Class<?>, RuleStats> stats = new HashMap<>();
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/StatsRecorder.java]: > Rename StatsRecorder to RuleStatsRecorder

doesn't quite look like a rename?

also, could you mention in the commit message why you rename the class?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/OptimizerStatsRecorder.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            }
+        }
+    }
+
+    void unexport(MBeanExporter exporter)
+    {
+        for (Class<?> rule : stats.keySet()) {
+            String name = ObjectNames.builder(PlanOptimizer.class)
+                    .withProperty("optimizer", rule.getSimpleName())
+                    .build();
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/OptimizerStatsRecorder.java]: Consider extracting object name building to a method to emphasize that export and unexport use the same naming strategy.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizerStats.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+package com.facebook.presto.sql.planner.optimizations;
+
+import io.airlift.stats.TimeDistribution;
+import org.weakref.jmx.Managed;
+import org.weakref.jmx.Nested;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class OptimizerStats
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizerStats.java]: Is it any different than rule stats?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StatsRecordingPlanOptimizer.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        }
+        catch (RuntimeException e) {
+            stats.recordFailure(this);
+            throw e;
+        }
+        stats.record(this, duration);
+        return result;
+    }
+
+    protected abstract PlanNode doOptimize(
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StatsRecordingPlanOptimizer.java]: Composition instead of inheritance? So that we leave possibility for an optimizer to pick up a base class?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/StatsRecorder.java (hunk 1 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/StatsRecorder.java]: also, why we don't simply use same stats objects for rules and optimizers?
If we want to distinguish them on the JMX level, we can do this with names. Now we have stats exported and stats recorder classes duplicated.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/StatsRecorder.java [hunk 5]
```diff
@@ -1,79 +0,0 @@
...
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-package com.facebook.presto.sql.planner;
-
-import com.facebook.presto.sql.planner.iterative.IterativeOptimizer;
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/StatsRecorder.java]: I think `RuleStatsRecorder` is not part of this commit (it should be)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/OptimizerStatsRecorder.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        requireNonNull(optimizer, "optimizer is null");
+        checkArgument(!optimizer.getClass().isAnonymousClass());
+        stats.put(optimizer.getClass(), new OptimizerStats());
+    }
+
+    public void record(PlanOptimizer optimizer, long nanos)
+    {
+        requireNonNull(optimizer, "optimizer is null");
+        stats.get(optimizer.getClass()).record(nanos);
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/OptimizerStatsRecorder.java]: add check?
`requireNonNull(stats.get(...), "optimizer X not registered")`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StatsRecordingPlanOptimizer.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        try {
+            long start = System.nanoTime();
+            result = doOptimize(plan, session, types, symbolAllocator, idAllocator);
+            duration = System.nanoTime() - start;
+        }
+        catch (RuntimeException e) {
+            stats.recordFailure(this);
+            throw e;
+        }
+        stats.record(this, duration);
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StatsRecordingPlanOptimizer.java]: Does recording duration of `0` modifies stats in case of failure (I recall 0 might be meaningful data point because there is count metrics in `TimeDistribution`).

Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StatsRecordingPlanOptimizer.java]: in case of failure, we won't be here. The fail exc is rethrown.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizerStats.java (hunk 3 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizerStats.java]: It does not have `hits`
